src/app/content/highlights/components/CardWrapper.tsx

```diff
@@ -55,20 +56,30 @@ const Wrapper = ({highlights, className, container, highlighter}: WrapperProps)
   };
 
   const resetTopOffset = () => {
-    element.current!.style.top = '0';
+    assertNotNull(element.current, 'element.current can\'t be null').style.top = '0';
   };
 
-  const updatePositions = React.useCallback(() => {
+  const getTopOffsetForHighlight = (highlight: Highlight) => {
+    if (topOffsets.has(highlight.id)) {
+      return assertDefined(topOffsets.get(highlight.id), 'this has to be defined');
+    } else {
+      const offset = assertDefined(
+        getHighlightTopOffset(container, highlight),
+        `Couldn't get top offset for highlight with an id: ${highlight.id}`
+      );
+      setTopOffsets((state) => new Map(state).set(highlight.id, offset));
```

Comment:
we're modifying state in a loop now, thats probably not great, i don't think this was here before, why are we adding it?

Reply:
Based on https://github.com/openstax/rex-web/pull/481#discussion_r395847393 I've wanted to memorize these values.
I think `setStates` should be merged and dispatched after the loop ends but I can't test it because servers are not working currently

Reply:
So I've checked it and these calls are dispatched after loop ends and are merged into single setState (ex. 9 calls will cause only one update of component after the state update).
If we want to avoid additional re-render (1 is from setting state by `updatePositions` and second is for those `setStates`) then I could make a `Map` outside of a component and clean it before component mounts in `useEffect`. What do you think?